src/keyring/providers/fortanixdsm.rs

```diff
@@ -0,0 +1,233 @@
+            },
+            public_key,
+        ))
+    }
+
+    fn sign(&self, msg: &[u8], hash_alg: DigestAlgorithm) -> Result<SignResponse, SignError> {
+        let req = SignRequest {
+            key: Some(self.descriptor.clone()),
+            data: Some(msg.to_owned().into()),
+            hash_alg,
+            hash: None,
+            mode: None,
+            deterministic_signature: None,
+        };
+        self.client.sign(&req).map_err(SignError::from_source)
+    }
+}
+
+impl Signer<EcdsaSignature> for SigningKey {
+    fn try_sign(&self, msg: &[u8]) -> Result<EcdsaSignature, SignError> {
+        assert_eq!(self.elliptic_curve, EllipticCurve::SecP256K1);
+        let resp = self.sign(msg, DigestAlgorithm::Sha256)?;
+        EcdsaSignature::from_der(&resp.signature)
+    }
+}
+
+impl Signer<Ed25519Signature> for SigningKey {
+    fn try_sign(&self, msg: &[u8]) -> Result<Ed25519Signature, SignError> {
+        assert_eq!(self.elliptic_curve, EllipticCurve::Ed25519);
+        let resp = self.sign(msg, DigestAlgorithm::Sha512)?;
+        Ed25519Signature::from_bytes(&resp.signature)
+    }
+}
+
+fn add_key(
+    registry: &mut chain::Registry,
+    config: &SigningKeyConfig,
+    client: Arc<SdkmsClient>,
+) -> Result<(), Error> {
+    let (signing_key, public_key) =
+        SigningKey::new(client, config.key.clone(), config.key_type.clone())?;
+
+    match config.key_type {
+        KeyType::Account => {
+            let signer = keyring::ecdsa::Signer::new(
+                SigningProvider::FortanixDsm,
+                public_key,
+                Box::new(signing_key),
+            );
+            for chain_id in &config.chain_ids {
+                registry.add_account_key(chain_id, signer.clone())?;
+            }
+        }
+        KeyType::Consensus => {
+            let signer = keyring::ed25519::Signer::new(
+                SigningProvider::FortanixDsm,
+                public_key,
+                Box::new(signing_key),
+            );
+            for chain_id in &config.chain_ids {
+                registry.add_consensus_key(chain_id, signer.clone())?;
+            }
+        }
+    }
+
+    Ok(())
+}
+
+// See RFC 8410 section 3
+const ED_25519_OID: ObjectIdentifier = ObjectIdentifier::new("1.3.101.112");
+
+struct Ed25519PublicKey(Ed25519);
+
+impl DecodePublicKey for Ed25519PublicKey {}
+
+impl<'a> TryFrom<SubjectPublicKeyInfo<'a>> for Ed25519PublicKey {
+    type Error = SpkiError;
+
+    fn try_from(spki: SubjectPublicKeyInfo<'_>) -> Result<Self, Self::Error> {
+        spki.algorithm.assert_algorithm_oid(ED_25519_OID)?;
+
+        if spki.algorithm.parameters != None {
+            // TODO: once/if https://github.com/RustCrypto/formats/issues/354 is addressed we should use that error variant.
+            return Err(SpkiError::KeyMalformed);
+        }
+
+        Ed25519::from_bytes(spki.subject_public_key)
+            .map_err(|_| SpkiError::KeyMalformed)
+            .map(Ed25519PublicKey)
+    }
+}
```

Comment:
Likewise most of this is available as [`ed25519::pkcs8::PublicKeyBytes`](https://docs.rs/ed25519/latest/ed25519/pkcs8/struct.PublicKeyBytes.html)